keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1210 @@
+     subfeature, or have each cloud provider have to maintain their
+     own set of `-ginkgo.skip` rules. It would be better if the e2e
+     tests themselves could just figure out, somehow, whether they
+     were running under a cloud provider that intends to implement the
+     feature they are testing, or a cloud provider that doesn't.
+
+  3. In some cases, because the existing tests were only run on
+     certain clouds, it is not clear what the expected semantics are
+     on other clouds. For example, since `IPMode: Proxy` load
+     balancers can't preserve the client source IP in the way that
+     `ExternalTrafficPolicy: Local` expects, should they refuse to
+     provision a load balancer at all, or should they provision a load
+     balancer that fails to preserve the source IP?
+
+This KEP proposes new additions to `service.Status.LoadBalancer` and
+`service.Status.Conditions` to allow cloud providers to better
+communicate the status of load balancer support and provisioning, and
+new guidelines on how cloud providers should handle load balancers for
+services that they cannot fully support.
+
+(Note: although I repeatedly refer to service load balancers as being
+implemented by "cloud providers" here, everything also applies to
+non-cloud-provider components load balancers via the
+[`LoadBalancerClass`] field.)
+
+[removal of in-tree cloud providers]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/
+[`LoadBalancerClass`]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/1959-service-lb-class-field/
+
+## Motivation
+
+### Goals
+
+- Allow cloud providers to indicate that they are working on
+  provisioning load balancer infrastructure, so that
+  users/operators/tests can distinguish the case of "it is taking a
+  while for the cloud to provision the load balancer" from "the cloud
+  has failed to provision a load balancer" and "there is no cloud
+  provider so load balancers don't work".
+
+- Allow cloud providers to indicate explicitly when they are unable to
+  provision a particular `LoadBalancer` service, and why.
+
+- Allow cloud providers to indicate when they have provided an
+  "imperfect" load balancer that the user may or may not consider to
+  be "good enough" (e.g., providing a single-stack load balancer for a
+  dual-stack Service).
+
+- Allow users and e2e tests to either prevent or reliably recognize
+  the case where the cloud provider has provisioned a load balancer
+  without support for a particular feature the user/test needs (even
+  when that is a feature the cloud provider doesn't know about).
+
+- Update at least [cloud-provider-gcp] and [cloud-provider-kind] to
+  fully support the new status reporting functionality, and encourage
+  other providers to follow.
+
+- Update the e2e load balancer tests to use the new functionality to
+  determine which tests to skip, so that any cloud provider that
```

Comment:
I agree on the spirit disagree on the goal, if we want to define a behavior that given Status X it should be a noop we do a test that do exactly that, but we should not do a test that starts asserting Z and given Status X or Y behave different